src/toil_scripts/batch_alignment/bwa_alignment.py

```diff
@@ -1,223 +1,254 @@
-
-
-def run_bwa(job, inputs, ids):
+    :param dict ids: FileStore IDs for shared inputs
     """
-    Aligns two fastqs into a BAMFILE via BWA
-
-    :param JobFunctionWrappingJob job: Passed by Toil automatically
-    :param Namespace inputs: Input arguments (see main)
-    :param list ids: list of FileStore IDs (R1, R2, reference inputs)
-    """
-    work_dir = job.fileStore.getLocalTempDir()
-    file_names = ['r1.fq.gz', 'r2.fq.gz', 'ref.fa', 'ref.fa.amb', 'ref.fa.ann',
-                  'ref.fa.bwt', 'ref.fa.pac', 'ref.fa.sa', 'ref.fa.fai']
-    if inputs.alt:
-        file_names.append('ref.fa.alt')
-
-    for fileStoreID, name in zip(ids, file_names):
-        job.fileStore.readGlobalFile(fileStoreID, os.path.join(work_dir, name))
-
-    rg = ''
-    if inputs.rg_line:
-        rg = inputs.rg_line
+    uuid, urls = sample
+    r1_url, r2_url = urls if len(urls) == 2 else urls[0], None
+    job.fileStore.logToMaster('Downloaded sample: {0}\nStarting BWA Run: {0}'.format(uuid))
+    ids['r1'] = job.addChildJobFn(download_url_job, r1_url, s3_key_path=inputs.ssec, disk=inputs.file_size).rv()
+    if r2_url:
+        ids['r2'] = job.addChildJobFn(download_url_job, r2_url, s3_key_path=inputs.ssec, disk=inputs.file_size).rv()
     else:
-        rg = "@RG\\tID:{0}".format(inputs.uuid) # '\' character is escaped so bwakit gets individual '\' and 't' characters 
-        for tag, info in zip(['LB', 'PL', 'PU', 'SM'], ['library', 'platform', 'program_unit', 'uuid']):
-            if hasattr(inputs, info):
-                rg = rg + '\\t{0}:{1}'.format(tag, getattr(inputs, info))
-
-    # BWA Options
-    opt_args = []
-    if inputs.sort:
-        opt_args.append('-s')
-    if inputs.trim_adapters:
-        opt_args.append('-a')
-    # Call: bwakit
-    parameters = (['-t', str(inputs.cores),
-                   '-R', rg] +
-                  opt_args +
-                  ['-o', '/data/aligned',
-                   '/data/ref.fa',
-                   '/data/r1.fq.gz',
-                   '/data/r2.fq.gz'])
-    outputs = {'aligned.aln.bam': inputs.mock_bam}
-
-    docker_call(tool='quay.io/ucsc_cgl/bwakit:0.7.12--528bb9bf73099a31e74a7f5e6e3f2e0a41da486e',
-                parameters=parameters, inputs=file_names, outputs=outputs, work_dir=work_dir)
-
-    # BWA insists on adding an `*.aln.sam` suffix, so rename the output file
-    output_file = os.path.join(work_dir, '{}.bam'.format(inputs.uuid))
-    os.rename(os.path.join(work_dir, 'aligned.aln.bam'),
-              output_file)
-
-    # Either write file to local output directory or upload to S3 cloud storage
-    job.fileStore.logToMaster('Aligned sample: {}'.format(inputs.uuid))
-
-    if inputs.output_dir.startswith('s3://'):
-        s3am_upload(output_file, inputs.output_dir, s3_key_path=inputs.ssec)
+        ids['r2'] = None
+    # Define and wire job functions
+    bam_id = job.wrapJobFn(run_bwakit, inputs.cores, ids['r1'], ids['r2'], ids['ref'], ids['fai'], ids['amb'],
```

Comment:
OK, squirmy baby drifted off to sleep. A variety of thoughts:

* `run_bwakit` now takes 19 parameters. This doesn't seem like a positive refactoring outcome to me, esp. since we're unpacking a dictionary to pass those 19 parameters. I understand the drive towards having just fileStore IDs passed around, but I think that having 9 fileStore IDs + many various parameters is too cumbersome.
* What's the rationale to decomposing `run_bwakit`, `s3am_upload_job`, `move_file_job` into child jobs? IMO, there's no real gain. They're not logically separate stages. You don't gain concurrency. I mean, maybe I'm being an idiot—and if I am, you should tell me—but isn't this always going to lead to the aligned BAM getting written to the filestore? If we're writing the file into S3 or to a local directory, this just doesn't make sense.
* Disk requirement should really be 3x `file_size` + index size, no?